src/lib/mina/test/test-contract.ts

```diff
@@ -0,0 +1,319 @@
+  },
+  callback: (input: {
+    accounts: Record<string, Mina.TestPublicKey>;
+    newAccounts: Record<string, Mina.TestPublicKey>;
+    contract: S;
+    Local: LocalBlockchain;
+  }) => TestInstruction[]
+): Promise<LocalBlockchain> {
+  // instance-independent setup: compile programs
+
+  offchainState?.setContractClass(Contract as any);
+  batchReducer?.setContractClass(Contract as any);
+
+  if (proofsEnabled) {
+    if (offchainState !== undefined) {
+      console.time('compile program');
+      await offchainState.compile();
+      console.timeEnd('compile program');
+    }
+    if (batchReducer !== undefined) {
+      console.time('compile reducer');
+      await batchReducer.compile();
+      console.timeEnd('compile reducer');
+    }
+    console.time('compile contract');
+    await Contract.compile();
+    console.timeEnd('compile contract');
+  }
+
+  // how to execute this test against a particular local Mina instance
+
+  async function execute(Local: LocalBlockchain) {
+    Mina.setActiveInstance(Local);
+
+    // set up accounts and connect contract to offchain state, reducer
+
+    let [sender, contractAccount] = Local.testAccounts;
+
+    let originalAccounts: Record<string, Mina.TestPublicKey> = {
+      sender,
+      contractAccount,
+    };
+    let i = 2;
+    let accounts: Record<string, Mina.TestPublicKey> = new Proxy(
+      originalAccounts,
+      {
+        get(accounts, name: string) {
+          if (name in accounts) return accounts[name];
+          let account = Local.testAccounts[i++];
+          assert(account !== undefined, 'ran out of test accounts');
+          accounts[name] = account;
+          return account;
+        },
+      }
+    );
+
+    let newAccounts: Record<string, Mina.TestPublicKey> = new Proxy(
+      {},
+      {
+        get(accounts, name: string) {
+          if (name in accounts) return newAccounts[name];
+          let account = Mina.TestPublicKey.random();
+          newAccounts[name] = account;
+          return account;
+        },
+      }
+    );
+
+    let contract = new Contract(contractAccount);
+    offchainState?.setContractInstance(contract as any);
+    batchReducer?.setContractInstance(contract as any);
+
+    // run test setup to return instructions
+    let instructions = callback({
+      accounts,
+      newAccounts,
+      contract: contract as S,
+      Local,
+    });
+
+    // deploy is the implicit first instruction (can be disabled with autoDeploy = false)
+    // TODO: figure out if the contract is already deployed on Mina instance,
+    // and only deploy if it's not
+    if (autoDeploy) instructions.unshift(deploy());
+
+    // run instructions
+    let spec = { localInstance: Local, contractClass: Contract };
+
+    for (let instruction of instructions) {
+      await runInstruction(spec, instruction);
+    }
+  }
+
+  // create local instance and execute test
+  // if proofsEnabled is 'both', run the test with AND without proofs
+
+  console.log();
```

Comment:
Is there a purpose for this empty `console.log`?

Reply:
I thought it makes the log output look better with the extra line break :D 
Not sure if we want to leave the testing framework as verbose as it is currently when we release it though, probably should have a `verbose` flag